### PR TITLE
Checkpoint refactoring - recognize checkpoints by operator instance name.

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
Use operator names rather than incidental order in checkpoints.
Get operator instances from executor (by name) rather than from lowered graph (by index).
Rationale:
- new graph doesn't store operator instances - that's the job of the executor
- order of operators in the graph is incidental and brittle

Additionally, OpCheckpoint is now only forward-declared in operator.h, because most operators are stateless and don't need to depend on OpCheckpoint definition.

### Tests:
Tests **adjusted** (not added)
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3979

### CODE REVIEW

File: dali/pipeline/operator/checkpointing/checkpoint.cc [hunk 1]
```diff
@@ -12,28 +12,33 @@
...
+  cpts_.emplace_back(std::move(instance_name));
+  return cpts_.size() - 1;
 }
 
-OpCheckpoint &Checkpoint::GetOpCheckpoint(OpNodeId id) {
-  DALI_ENFORCE_VALID_INDEX(id, cpts_.size());
-  return cpts_[id];
+OpCheckpoint &Checkpoint::GetOpCheckpoint(int idx) {
+  DALI_ENFORCE_VALID_INDEX(idx, cpts_.size());
+  return cpts_[idx];
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operator/checkpointing/checkpoint.cc:36]: OK, there's a much bigger problem here. It needs urgent fixing.

[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operator/checkpointing/checkpoint.cc:36]: Well, here the executor is expected to remember the order in which the operators were added to the checkpoint. Still, there was a bug elsewhere and I've already posted a fix.

[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operator/checkpointing/checkpoint.cc:36]: It's a question of practicality - current executor uses indices. I could add a name->index map and a lookup by name function, but I can still do it later, if it's necessary. Note that when restoring the checkpoint, I reversed the old logic and now restore the operators' state in the order in which they appear in the checkpoint, getting operators from the executor by name.

**Review by szkarpinski — APPROVED** [review]

File: dali/pipeline/executor/executor_impl.cc [hunk 2]
```diff
@@ -748,8 +751,15 @@ Checkpoint &Executor<WorkspacePolicy, QueuePolicy>::GetCurrentCheckpoint() {
...
-    graph_->Node(i).op->RestoreState(cpt.GetOpCheckpoint(i));
+
+  for (int i = 0, n = cpt.NumOp(); i < n; i++) {
+    const auto &op_cpt = cpt.GetOpCheckpoint(i);
+    auto *op = GetOperator(op_cpt.OperatorName());
+    DALI_ENFORCE(op != nullptr, make_string(
+        "Operator with name \"", op_cpt.OperatorName(), "\" "
+        "from the checkpoint not found in the graph."));
+    op->RestoreState(op_cpt);
+  }
```
Comment (stiepan) [inline dali/pipeline/executor/executor_impl.cc:762]: There's a slight change in what checkpoints we accept now, namely, if the pipeline has some extra ops that are not present in the checkpoint, we will silently accept it and start with some of the ops, potentially, not restored.

How about iterating here over the graph and getting the OpCheckpoints by name? Together with the deserialization part we would check inclusions in both direction.

The separate issue is whether we may want behaviour of this kind, for example, explictly allowing for no OpCheckpoint for stateless operators. But that deserves a follow-up, what I am worried about here is silently loosening the validation.

Comment (mzient) [reply dali/pipeline/executor/executor_impl.cc:763]: Added a check that the number of operators matches.

**Review by stiepan — APPROVED** [review]

**Review by szkarpinski — APPROVED** [review]

**Pull request merged**